### ISSUE TITLE
can not set device

### ISSUE THREAD

**Issue by qianqing13579** [ROCmSoftwarePlatform/AMDMIGraphX#876]
I use set_device() to run a model on a device,but it does not work!

**Comment by causten** [comment]
Can you show where set_device() comes from?

**Comment by qianqing13579** [comment]
The set_decice function is in src/targets/gpu/hip.cpp

**Comment by qianqing13579** [comment]
@causten ,I have read the source code,and I find that the gpu::context set the  device_id to 0,so the model will always run in 0 device

**Comment by pfultz2** [comment]
> I have read the source code,and I find that the gpu::context set the device_id to 0,so the model will always run in 0 device

Ah yes, we probably need a way to specify a different device when loading/running a model.

**Comment by kahmed10** [comment]
@qianqing13579 we don't have a way to set GPUs in MIGraphx currently, but for now if you have a multi-GPU machine, you can compile and run MIGraphX on a specific GPU by setting the `ROCR_VISIBLE_DEVICES` environment variable.

For example:
```
ROCR_VISIBLE_DEVICES=1 ./migraphx-driver run --model alexnet
```
This will pin MIGraphX to run on GPU 1.

**Comment by qianqing13579** [comment]
@kahmed10 ,I have solved the problem. I modified the src/targets/gpu/target.cpp and src/program.cpp:
![2021-09-18_083650](https://user-images.githubusercontent.com/8730200/133866099-bfe256ab-cca2-4c5e-86c9-9672641eeef6.jpg)
![2021-09-18_083812](https://user-images.githubusercontent.com/8730200/133866105-3a530d31-4d73-4132-894c-8dbe7a3a0352.jpg)

**Comment by kahmed10** [comment]
Great. Would you like to open an PR for this change?

**Comment by qianqing13579** [comment]
@kahmed10 ,ok

**Comment by qianqing13579** [comment]
@kahmed10 ,do I need to create a new branch?

**Comment by kahmed10** [comment]
> @kahmed10 ,do I need to create a new branch?
I believe you can fork the repository and then create the PR. I don't think you will be able to create any new branches in the official repo.

**Comment by qianqing13579** [comment]
@kahmed10, @pfultz2 , I have made a commit for this change.the link:https://github.com/qianqing13579/AMDMIGraphX/commit/94df31b9a62398405d6c52f098923d607d470dce

**Comment by qianqing13579** [comment]
@kahmed10 , MIGraphX does not need to have this functionality. The functionality is developed by users in  theirs program. the demo is as follows:
```
// read device id from the configuration file
FileStorage configurationFile;
configurationFile.open("myconfig.xml", FileStorage::READ);
FileNode netNode = configurationFile["mymodel"];
int device_id=(int)netNode["device_id"];

// set device id
migraphx::compile_options options;
options.device_id=device_id;
net.compile(migraphx::gpu::target{},options);
```

**Comment by zixianwang2022** [comment]
Hi, @kahmed10 @qianqing13579 

Is allocating model to a specific gpu implemented now? I am trying to run data-parallel that needs each GPU hosting 1 complete instance of model.

**Comment by pfultz2** [comment]
Yes, this was fixed by #1561.
